components/tidb_query/src/batch/runner.rs

```diff
@@ -374,14 +374,12 @@ impl<SS: 'static> BatchExecutorsRunner<SS> {
         let mut warnings = self.config.new_eval_warnings();
 
         loop {
-            // return (is_drained, EvalWarnings, record_cnt, Chunk)
             let mut chunk = Chunk::default();
-            let (is_drained, mut result_warnings, record_cnt) =
-                self.internal_handle_request(batch_size, &mut chunk)?;
+            let (is_drained, record_len) =
+                // return (is_drained, record_len)
+                self.internal_handle_request(batch_size, &mut chunk, &mut warnings)?;
 
-            warnings.merge(&mut result_warnings);
-
-            if record_cnt > 0 {
+            if record_len > 0 {
                 chunks.push(chunk);
```

Comment:
so why not make `internal_handle_request` push chunk directly?

Reply:
> so why not make `internal_handle_request` push chunk directly?

In streaming mode, only one `chunk` variable is created, and `internal_handle_request` will push data to it. And in non-streaming mode, a `Vec<Chunk>` is used. So only in non-streaming mode, we need to use `chunks.push(chunk);`.